src/main/java/com/google/devtools/build/lib/remote/RemoteOutputChecker.java

```diff
@@ -363,4 +359,35 @@ public void maybeInvalidateSkyframeValues(MemoizingEvaluator memoizingEvaluator)
           });
     }
   }
+
+  /**
+   * A specialized concurrent trie that stores paths of artifacts and allows checking whether a
+   * given path is contained in (in the case of a tree artifact) or exactly matches (in any other
+   * case) an artifact in the trie.
+   */
+  private static final class ConcurrentArtifactPathTrie {
+    // Invariant: no path in this set is a prefix of another path.
+    private final ConcurrentSkipListSet<PathFragment> paths = new ConcurrentSkipListSet<>();
+
+    /**
+     * Adds the given {@link ActionInput} to the trie.
+     *
+     * <p>The caller must ensure that no object's path passed to this method is a prefix of any
+     * previously added object's path. Bazel enforces this for non-aggregate artifacts. Callers must
+     * not pass in {@link TreeFileArtifact}s (which have exec paths that are prefixes of their
+     * parent tree artifact's exec path) or non-Artifact {@link ActionInput}s that violate this
+     * invariant.
```

Comment:
What happens in the presence of action templates (see `ActionTemplate.java`)? They expand into actions whose outputs are individual `TreeFileArtifacts` (\with `isChildOfDeclaredDirectory() == false`). But I'm not sure whether `TopLevelArtifactHelper.getAllArtifactsToBuild` will surface the `TreeFileArtifacts`, the `TreeArtifact` they belong to, or both (which would be bad for this invariant).

Reply:
I ran CI with a `checkArgument` for `input` not being a `TreeFileArtifact` and it passed - we do have coverage for this case since https://github.com/bazelbuild/bazel/commit/00bb86b01397d3ad6f3794077fea2958c06d817b.

From what I can tell, we only ever pass aggregate artifacts to this method and the logic in `ImportantOutputHandler` handles the expansion.

Reply:
Thanks for checking. Let's leave the precondition check in, then? If this invariant is broken in the future I'd rather see a crash than a silent bug.